scripts/setup-grafana.sh

```diff
@@ -0,0 +1,122 @@
+#!/bin/bash
+
+if [ "$#" -ne 2 ]; then
+    echo "Usage: $0 <instance> <project>"
+    echo "Error: Both 'instance' and 'project' arguments are required."
+    exit 1
+fi
+
+INSTANCE=$1
+PROJECT=$2
+SERVER_NAME="$(lxc info | awk '{if ($1 == "server_name:") print $2}')"
+IS_LXD_CLUSTERED=$(lxc info | grep "server_clustered:" | grep "false")
+if [ -z "$IS_LXD_CLUSTERED" ]; then
+    echo "Error: LXD is clustered, this script only works for single node installations."
+    echo "See https://documentation.ubuntu.com/lxd/en/latest/metrics/ for more information."
+    exit 1
+fi
+
+CONTAINER_IP="$(lxc list "${INSTANCE}" --project "${PROJECT}" -c 4 -f csv | awk '{print $1}')"
+CONTAINER_UPLINK_IP=$(echo "$CONTAINER_IP" | cut -d "." -f1,2,3)".1"
+echo "Found container IP as '$CONTAINER_IP' and uplink as '$CONTAINER_UPLINK_IP'"
+
+set -e
+set -x
+
+# upload server.crt to container
+lxc info | sed -n "/BEGIN CERTIFICATE/,/END CERTIFICATE/ s/^\s*// p" | lxc file push - "$INSTANCE"/root/server.crt --project="$PROJECT"
+
+# install and configure grafana and prometheus in container
+lxc exec "$INSTANCE" --project="$PROJECT" bash <<EOF
+set -x
+set -e
+# install grafana and prometheus
+apt-get update
+apt-get install -y software-properties-common wget prometheus
+mkdir -p /etc/apt/keyrings/
+wget -q -O - https://apt.grafana.com/gpg.key | gpg --dearmor | tee /etc/apt/keyrings/grafana.gpg > /dev/null
+echo "deb [signed-by=/etc/apt/keyrings/grafana.gpg] https://apt.grafana.com stable main" | tee -a /etc/apt/sources.list.d/grafana.list
+apt-get update
+apt-get install -y grafana=11.6.0 loki=3.4.2 promtail=3.4.2
+systemctl daemon-reload
+systemctl enable --now grafana-server.service
+systemctl enable --now loki
+systemctl enable --now promtail
+
+# generate ssl key for grafana to serve via https
+openssl req -x509 -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -sha384 -keyout /etc/grafana/grafana.key -out /etc/grafana/grafana.crt -days 365 -nodes -subj "/CN=metrics.local"
+chown grafana:grafana /etc/grafana/grafana.crt
+chown grafana:grafana /etc/grafana/grafana.key
+chmod 0640 /etc/grafana/grafana.key /etc/grafana/grafana.crt
+sed -i "s#;protocol = http#protocol = https#" /etc/grafana/grafana.ini
+cat <<EOT > /etc/grafana/provisioning/datasources/lxd-sources.yaml
+apiVersion: 1
+
+datasources:
+  - name: prometheus
+    type: prometheus
+    access: proxy
+    url: http://$CONTAINER_IP:9090
+  - name: loki
+    type: loki
+    access: proxy
+    url: http://$CONTAINER_IP:3100
+EOT
+systemctl restart grafana-server
+
+# generate certs for prometheus
+openssl req -x509 -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -sha384 -keyout metrics.key -nodes -out metrics.crt -days 3650 -subj "/CN=metrics.local"
+mkdir /etc/prometheus/tls
+mv metrics.* /etc/prometheus/tls/
+mv server.crt /etc/prometheus/tls/
+chown -R prometheus:root /etc/prometheus/tls
```

Comment:
```suggestion
chown root:prometheus /etc/prometheus/tls/metrics.key
```
This way, the prometheus daemon can read the private key to use it but cannot change it.

Reply:
I think this didn't work for me earlier, I wouldn't get metrics with this change.

Reply:
Yes, this breaks the setup.

Reply:
Hmm, could you share the output of `namei -l /etc/prometheus/tls/metrics.key`, please?

It should look like this:
```
# namei -l /etc/prometheus/tls/metrics.key
f: /etc/prometheus/tls/metrics.key
drwxr-xr-x root root       /
drwxr-xr-x root root       etc
drwxr-x--- root prometheus prometheus
drwxr-x--- root prometheus tls
-rw-r----- root prometheus metrics.key
```

Reply:
What is the problem with owning the whole directory and contents to the prometheus user and root group? Why change it in the first place to root user and prometheus group? The former works well already :)

Reply:
Indeed, it's not about functional or not but more a question of least privileges.

Reply:
I think the existing `prometheus:root` is giving lower permissions than the suggested `root:prometheus`.

Reply:
With your method, the prometheus daemon can alter any file under `/etc/prometheus/tls` which IMHO is undesirable. It only needs read access to all files which is why I suggested tweaking the permissions/ownership.

I just ran the script with my suggested changes and it works. Here's the resulting `namei` output:

```
root@grafana:~# namei -l /etc/prometheus/tls/metrics.key 
f: /etc/prometheus/tls/metrics.key
drwxr-xr-x root root       /
drwxr-xr-x root root       etc
drwxr-xr-x root root       prometheus
drwxr-xr-x root root       tls
-rw-r----- root prometheus metrics.key
```

The bit that was broken is when the `server.crt` file was pushed into the container. I'll propose a change to address this.